<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка" data-toc-modified-id="Подготовка-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Загрузим-библиотеки" data-toc-modified-id="Загрузим-библиотеки-1.0.1"><span class="toc-item-num">1.0.1&nbsp;&nbsp;</span>Загрузим библиотеки</a></span></li><li><span><a href="#Загружаем-данные" data-toc-modified-id="Загружаем-данные-1.0.2"><span class="toc-item-num">1.0.2&nbsp;&nbsp;</span>Загружаем данные</a></span></li><li><span><a href="#Изучаем-данные" data-toc-modified-id="Изучаем-данные-1.0.3"><span class="toc-item-num">1.0.3&nbsp;&nbsp;</span>Изучаем данные</a></span></li><li><span><a href="#Изучим-столбец-с-целевым-признаком" data-toc-modified-id="Изучим-столбец-с-целевым-признаком-1.0.4"><span class="toc-item-num">1.0.4&nbsp;&nbsp;</span>Изучим столбец с целевым признаком</a></span></li><li><span><a href="#Вывод:" data-toc-modified-id="Вывод:-1.0.5"><span class="toc-item-num">1.0.5&nbsp;&nbsp;</span>Вывод:</a></span></li><li><span><a href="#Токенизируем-и-лемматизируем-столбец-text-и-добавим-данные-в-столбец-text_final" data-toc-modified-id="Токенизируем-и-лемматизируем-столбец-text-и-добавим-данные-в-столбец-text_final-1.0.6"><span class="toc-item-num">1.0.6&nbsp;&nbsp;</span>Токенизируем и лемматизируем столбец text и добавим данные в столбец text_final</a></span></li><li><span><a href="#Выведем-одну-строку-для-последующего-контроля" data-toc-modified-id="Выведем-одну-строку-для-последующего-контроля-1.0.7"><span class="toc-item-num">1.0.7&nbsp;&nbsp;</span>Выведем одну строку для последующего контроля</a></span></li><li><span><a href="#Запускаем-преобразование-с-контролем-времени-работы" data-toc-modified-id="Запускаем-преобразование-с-контролем-времени-работы-1.0.8"><span class="toc-item-num">1.0.8&nbsp;&nbsp;</span>Запускаем преобразование с контролем времени работы</a></span></li><li><span><a href="#Проверим-преобразование" data-toc-modified-id="Проверим-преобразование-1.0.9"><span class="toc-item-num">1.0.9&nbsp;&nbsp;</span>Проверим преобразование</a></span></li><li><span><a href="#Создадим-признаки-и-целевой-признак" data-toc-modified-id="Создадим-признаки-и-целевой-признак-1.0.10"><span class="toc-item-num">1.0.10&nbsp;&nbsp;</span>Создадим признаки и целевой признак</a></span></li><li><span><a href="#Разобъем-на-выборки-тренировочную-и-тестовую" data-toc-modified-id="Разобъем-на-выборки-тренировочную-и-тестовую-1.0.11"><span class="toc-item-num">1.0.11&nbsp;&nbsp;</span>Разобъем на выборки тренировочную и тестовую</a></span></li><li><span><a href="#Проверим-распределения-классов-в-выборках" data-toc-modified-id="Проверим-распределения-классов-в-выборках-1.0.12"><span class="toc-item-num">1.0.12&nbsp;&nbsp;</span>Проверим распределения классов в выборках</a></span></li><li><span><a href="#Создадим-корпус" data-toc-modified-id="Создадим-корпус-1.0.13"><span class="toc-item-num">1.0.13&nbsp;&nbsp;</span>Создадим корпус</a></span></li><li><span><a href="#Укажем-стоп-слова" data-toc-modified-id="Укажем-стоп-слова-1.0.14"><span class="toc-item-num">1.0.14&nbsp;&nbsp;</span>Укажем стоп-слова</a></span></li><li><span><a href="#Векторизируем-наши-признаки-с-учетом-стоп-слов" data-toc-modified-id="Векторизируем-наши-признаки-с-учетом-стоп-слов-1.0.15"><span class="toc-item-num">1.0.15&nbsp;&nbsp;</span>Векторизируем наши признаки с учетом стоп-слов</a></span><ul class="toc-item"><li><span><a href="#TF-IDF" data-toc-modified-id="TF-IDF-1.0.15.1"><span class="toc-item-num">1.0.15.1&nbsp;&nbsp;</span>TF-IDF</a></span></li></ul></li><li><span><a href="#Проверим-данные-после-векторизации" data-toc-modified-id="Проверим-данные-после-векторизации-1.0.16"><span class="toc-item-num">1.0.16&nbsp;&nbsp;</span>Проверим данные после векторизации</a></span></li><li><span><a href="#Вывод:" data-toc-modified-id="Вывод:-1.0.17"><span class="toc-item-num">1.0.17&nbsp;&nbsp;</span>Вывод:</a></span></li></ul></li></ul></li><li><span><a href="#Обучение" data-toc-modified-id="Обучение-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обучение</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Обучим-следующие-модели:" data-toc-modified-id="Обучим-следующие-модели:-2.0.1"><span class="toc-item-num">2.0.1&nbsp;&nbsp;</span>Обучим следующие модели:</a></span></li><li><span><a href="#Обучим-LogisticRegression-и-вычислим-F1-score-TF-IDF" data-toc-modified-id="Обучим-LogisticRegression-и-вычислим-F1-score-TF-IDF-2.0.2"><span class="toc-item-num">2.0.2&nbsp;&nbsp;</span>Обучим LogisticRegression и вычислим F1-score TF-IDF</a></span></li><li><span><a href="#Внесем-данные-в-таблицу-results" data-toc-modified-id="Внесем-данные-в-таблицу-results-2.0.3"><span class="toc-item-num">2.0.3&nbsp;&nbsp;</span>Внесем данные в таблицу results</a></span></li><li><span><a href="#Обучим-RandomForestClassifier-и--вычислим-F1-score-TF-IDF" data-toc-modified-id="Обучим-RandomForestClassifier-и--вычислим-F1-score-TF-IDF-2.0.4"><span class="toc-item-num">2.0.4&nbsp;&nbsp;</span>Обучим RandomForestClassifier и  вычислим F1-score TF-IDF</a></span></li><li><span><a href="#Обучим-DecisionTreeClassifier-и-вычислим-F1-score-TF-IDF" data-toc-modified-id="Обучим-DecisionTreeClassifier-и-вычислим-F1-score-TF-IDF-2.0.5"><span class="toc-item-num">2.0.5&nbsp;&nbsp;</span>Обучим DecisionTreeClassifier и вычислим F1-score TF-IDF</a></span></li></ul></li></ul></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Выводы</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Проект для «Викишоп»

Интернет-магазин «Викишоп» запускает новый сервис. Теперь пользователи могут редактировать и дополнять описания товаров, как в вики-сообществах. То есть клиенты предлагают свои правки и комментируют изменения других. Магазину нужен инструмент, который будет искать токсичные комментарии и отправлять их на модерацию. 

Обучите модель классифицировать комментарии на позитивные и негативные. В вашем распоряжении набор данных с разметкой о токсичности правок.

Постройте модель со значением метрики качества *F1* не меньше 0.75. 

**Инструкция по выполнению проекта**

1. Загрузите и подготовьте данные.
2. Обучите разные модели. 
3. Сделайте выводы.

Для выполнения проекта применять *BERT* необязательно, но вы можете попробовать.

**Описание данных**

Данные находятся в файле `toxic_comments.csv`. Столбец *text* в нём содержит текст комментария, а *toxic* — целевой признак.

## Подготовка

#### Загрузим библиотеки

In [1]:
import pandas as pd
import numpy as np

from IPython.display import display
import warnings
warnings.filterwarnings('ignore')

import csv
import nltk
from nltk.corpus import stopwords as nltk_stopwords
from sklearn.feature_extraction.text import TfidfVectorizer

from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import f1_score
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier

import lightgbm as lgb
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.metrics import f1_score

from nltk.stem import WordNetLemmatizer
# from pymystem3 import Mystem
m = WordNetLemmatizer() 
import re
from tqdm import notebook
notebook.tqdm.pandas() 

#### Загружаем данные

In [2]:
data = pd.read_csv('/datasets/toxic_comments.csv')

#### Изучаем данные

In [3]:
display(data.head())
display(data.info())

text  toxic
0  Explanation\nWhy the edits made under my usern...      0
1  D'aww! He matches this background colour I'm s...      0
2  Hey man, I'm really not trying to edit war. It...      0
3  "\nMore\nI can't make any real suggestions on ...      0
4  You, sir, are my hero. Any chance you remember...      0

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 159571 entries, 0 to 159570
Data columns (total 2 columns):
text     159571 non-null object
toxic    159571 non-null int64
dtypes: int64(1), object(1)
memory usage: 2.4+ MB


None

#### Изучим столбец с целевым признаком

In [4]:
display(data['toxic'].value_counts())
data['toxic'].value_counts(normalize=True).map('{:.2%}'.format)

0    143346
1     16225
Name: toxic, dtype: int64

0    89.83%
1    10.17%
Name: toxic, dtype: object

#### Вывод:
* Данные загружены, состоят из 159571 строк и 2 столбцов
* Столбец toxic является целевым признаком с выраженым дисбалансом классов

#### Токенизируем и лемматизируем столбец text и добавим данные в столбец text_final

In [5]:
def text_tokenize(text):
    tokenized = nltk.word_tokenize(text)
    joined = ' '.join(tokenized)
    text_only = re.sub(r'[^a-zA-Z]', ' ', joined)
    final = ' '.join(text_only.split())
    x =  "".join(m.lemmatize(final))
    return x

#### Выведем одну строку для последующего контроля

In [6]:
display(data['text'][0])

"Explanation\nWhy the edits made under my username Hardcore Metallica Fan were reverted? They weren't vandalisms, just closure on some GAs after I voted at New York Dolls FAC. And please don't remove the template from the talk page since I'm retired now.89.205.38.27"

#### Запускаем преобразование с контролем времени работы

In [7]:
notebook.tqdm.pandas() 
data['text_final'] = data['text'].progress_apply(text_tokenize)

#### Проверим преобразование

In [8]:
display(data['text_final'][0])

'Explanation Why the edits made under my username Hardcore Metallica Fan were reverted They were n t vandalisms just closure on some GAs after I voted at New York Dolls FAC And please do n t remove the template from the talk page since I m retired now'

#### Создадим признаки и целевой признак

In [9]:
features = data['text_final']
target = data['toxic']

#### Разобъем на выборки тренировочную и тестовую

In [10]:
features_train, features_test, target_train, target_test = train_test_split(features, target, random_state=12345, 
                                                                            stratify=target)
print(features_train.shape)
print(target_train.shape)
print(features_test.shape)
print(target_test.shape)

(119678,)
(119678,)
(39893,)
(39893,)


#### Проверим распределения классов в выборках

In [11]:
print('Распределение классов в обучающей выборке:')
print(target_train.value_counts()[0] / target_train.value_counts().sum())
print(target_train.value_counts()[1] / target_train.value_counts().sum())
print()
print('Распределение классов в тестовой выборке:')
print(target_test.value_counts()[0] / target_test.value_counts().sum())
print(target_test.value_counts()[1] / target_test.value_counts().sum())

Распределение классов в обучающей выборке:
0.8983188221728304
0.10168117782716957

Распределение классов в тестовой выборке:
0.8983280274734916
0.1016719725265084


#### Создадим корпус

In [12]:
corpus_train = features_train.values.astype('U')
corpus_test = features_test.values.astype('U')

#### Укажем стоп-слова

In [13]:
nltk.download('stopwords')
stopwords = set(nltk_stopwords.words('english'))
display(len(stopwords))

[nltk_data] Downloading package stopwords to /home/jovyan/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


179

#### Векторизируем наши признаки с учетом стоп-слов

##### TF-IDF

In [14]:
tf_idf = TfidfVectorizer(stop_words=stopwords)

In [15]:
%%time
features_train_data_tf_idf = tf_idf.fit_transform(corpus_train)
features_test_data_tf_idf = tf_idf.transform(corpus_test)

CPU times: user 12.8 s, sys: 113 ms, total: 12.9 s
Wall time: 13 s


#### Проверим данные после векторизации

In [16]:
print('TF-IDF train')
print(features_train_data_tf_idf.shape)
print(features_train_data_tf_idf[:5].toarray())
print()
print('TF-IDF test')
print(features_test_data_tf_idf.shape)
print(features_test_data_tf_idf[:5].toarray())

TF-IDF train
(119678, 142074)
[[0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]]

TF-IDF test
(39893, 142074)
[[0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]]


#### Вывод:
* Токенизировали и лемматизировали столбец text и добавили данные в столбец text_final
* Провели оценку важности слов определением величины TF-IDF
* Проверили данные после векторизации

## Обучение

#### Обучим следующие модели:
* LogisticRegression
* RandomForestClassifier
* DecisionTreeClassifier.

#### Обучим LogisticRegression и вычислим F1-score TF-IDF

In [17]:
%%time
model = LogisticRegression(max_iter=500, random_state=12345, class_weight='balanced')
model.fit(features_train_data_tf_idf, target_train)
predictions = model.predict(features_test_data_tf_idf)
    
test_f1_score = f1_score(target_test, predictions)

CPU times: user 7.3 s, sys: 4.79 s, total: 12.1 s
Wall time: 12.1 s


#### Внесем данные в таблицу results

In [18]:
results = pd.DataFrame({'Preprocessing model' : ['TF-IDF'],
                        'Learning model' : ['LogisticRegression'], 
                        'Test f1 score' : round(test_f1_score, 2)})

#### Обучим RandomForestClassifier и  вычислим F1-score TF-IDF

In [19]:
%%time
model = RandomForestClassifier(random_state=12345, class_weight = 'balanced')
model.fit(features_train_data_tf_idf, target_train)
predictions = model.predict(features_test_data_tf_idf)
    
test_f1_score = f1_score(target_test, predictions)

CPU times: user 1min 9s, sys: 0 ns, total: 1min 9s
Wall time: 1min 12s


In [20]:
results = results.append({'Preprocessing model' : 'TF-IDF',
                        'Learning model' : 'RandomForestClassifier', 
                          'Test f1 score' : round(test_f1_score, 2)}, ignore_index=True)

#### Обучим DecisionTreeClassifier и вычислим F1-score TF-IDF

In [21]:
%%time
model = DecisionTreeClassifier(random_state=12345, class_weight = 'balanced')
model.fit(features_train_data_tf_idf, target_train)
predictions = model.predict(features_test_data_tf_idf)
    
test_f1_score = f1_score(target_test, predictions)

CPU times: user 3min 4s, sys: 2.48 ms, total: 3min 4s
Wall time: 3min 6s


In [22]:
results = results.append({'Preprocessing model' : 'TF-IDF',
                        'Learning model' : 'DecisionTreeClassifier',
                          'Test f1 score' : round(test_f1_score, 2)}, ignore_index=True)

In [23]:
display(results)

Preprocessing model          Learning model  Test f1 score
0              TF-IDF      LogisticRegression           0.76
1              TF-IDF  RandomForestClassifier           0.59
2              TF-IDF  DecisionTreeClassifier           0.66

## Выводы

* Данные загружены, состоят из 159571 строк и 2 столбцов
* Столбец toxic является целевым признаком с выраженым дисбалансом классов
* Токенизировали и лемматизировали столбец text и добавим данные в столбец text_final
* Провели оценку важности слов определением величины TF-IDF
* Проверили данные после векторизации
* Обучили следующие модели:
   * LogisticRegression
   * RandomForestClassifier
   * DecisionTreeClassifier
* Лучший результат метрики F1-score **0.76** показала модель LogisticRegression
* Другие модели не прошли по условию **F1-score не меньше 0.75** и время обучения и предсказания намного выше чем у модели LogisticRegression
* Модель **LogisticRegression** можно рекомендовать для использования в качестве инструмента, который будет искать токсичные комментарии и отправлять их на модерацию.